system/include/simd128.h

```diff
@@ -0,0 +1,729 @@
+/*
+WebAssembly SIMD128 Intrinsics
+*/
+
+#include <stdint.h>
+
+// User-facing type
+typedef int32_t v128_t __attribute__((vector_size(16), __aligned__(16)));
+
+// Internal types determined by clang builtin definitions
+typedef int32_t __v128_u __attribute__((__vector_size__(16), __aligned__(1)));
+typedef char  __i8x16 __attribute__((__vector_size__(16), __aligned__(16)));
+typedef unsigned char __u8x16 __attribute__((__vector_size__(16), __aligned__(16)));
+typedef short __i16x8 __attribute__((__vector_size__(16), __aligned__(16)));
+typedef unsigned short __u16x8 __attribute__((__vector_size__(16), __aligned__(16)));
+typedef int __i32x4 __attribute__((__vector_size__(16), __aligned__(16)));
+typedef unsigned int __u32x4 __attribute__((__vector_size__(16), __aligned__(16)));
+typedef long long __i64x2 __attribute__((__vector_size__(16), __aligned__(16)));
+typedef unsigned long long __u64x2 __attribute__((__vector_size__(16), __aligned__(16)));
+typedef float __f32x4 __attribute__((__vector_size__(16), __aligned__(16)));
+typedef double __f64x2 __attribute__((__vector_size__(16), __aligned__(16)));
+
+#define __DEFAULT_FN_ATTRS __attribute__((__always_inline__, __nodebug__, __target__("simd128"), __min_vector_width__(128)))
+
+// v128 wasm_v128_load(void* mem)
+static __inline__ v128_t __DEFAULT_FN_ATTRS wasm_v128_load(const void* __mem) {
+  struct __wasm_v128_load_struct {
+    __v128_u __v;
+  } __attribute__((__packed__, __may_alias__));
+  return ((const struct __wasm_v128_load_struct*)__mem)->__v;
+}
+
+// wasm_v128_store(void* mem, v128 a)
+static __inline__ void __DEFAULT_FN_ATTRS wasm_v128_store(void* __mem, v128_t __a) {
+  struct __wasm_v128_store_struct {
+    __v128_u __v;
+  } __attribute__((__packed__, __may_alias__));
+  ((struct __wasm_v128_store_struct*)__mem)->__v = __a;
+}
+
+// v128 wasm_v128_constant(...)
+static __inline__ v128_t __DEFAULT_FN_ATTRS wasm_v128_const(int8_t c0, int8_t c1, int8_t c2,
```

Comment:
We need multiple versions of this intrinsic function for different data types (e.g. 4 floats, 2 doubles, 4 `int32_t`s)

Reply:
Those operations can be implemented on top of this intrinsic, so we don't strictly need them.

Reply:
You may not need them with current Clang implementation, but if another compiler implements them with requirement that values must be known at compile-time (e.g. like Clang requires for `__builtin_shufflevector`), it won't be possible (the only way to reinterpret `float` as integer data that doesn't involve a UB is `memcpy`).

Reply:
I think there are two problems here.

The first problem is that this implementation does not statically reject runtime arguments. 

@tlively I think this should have a test that errors if the arguments to `wasm_v128_constant` are not compile-time constants, and that passes otherwise:

```c++
int8_t a = 0;
v128 b = wasm_v128_constant(a, a, ..., a); // ERROR: a is not a compile-time constant

constexpr int8_t a = 0;
v128 b = wasm_v128_constant(a, a, ..., a); // OK
```

That is, `wasm_v128_constant` supporting arguments that are run-time values is a clang-specific compiler extension that should IMO be opt-in (and not enabled by default). 

---

The second problem is: how does one implement a wrapper over `wasm_v128_const` that constructs a vector from 4 floats or similar. In C++, you would write this as follows: 

```c++
constexpr v128 wasm_v128_float(float a, float b, float c, float d) {
    if constexpr(false) { throw "can only be invoked at compile-time"; }
    using A = array<int8_t, 4>;
    auto aN = std::bit_cast<A>(a);
    ...
    auto dN = std::bit_cast<A>(d);
    return wasm_v128_const(aN[0], aN[1], ..., dN[3]);
}
```

In C, I don't think one can implement such a wrapper. People using C would need to convert the floats to `int8_t` by hand before calling the `wasm_v128_const` intrinsic. 

While this is bad, I am not sure whether adding more `wasm_v128_const_...` intrinsics for these operations is the right way to fix that (e.g. as opposed to fixing the underlying problem).

Reply:
Both good points. The only way I know of to enforce immediate arguments is by creating a new clang builtin function, so I will have to do a bit of extra LLVM work for that. I think adding new intrinsics for each lane type is also reasonable. @gnzlbg, what do you mean by "fixing the underlying problem" instead?

Reply:
> what do you mean by "fixing the underlying problem" instead?

The underlying problem is that C, as opposed to C++, has no way to perform a bitcast from a constant expression to another constant expression. Adding, e.g., a `wasm_v128_const_float` intrinsics, does not solve this problem. To solve this underlying problem, we could add an intrinsic similar to C++'s `std::bitcast` to C. That would allow implementing `wasm_v128_const_float` as follows:

```c
#define wasm_v128_const_float(a, b, c, d) \
  wasm_v128_const(bitcast<int8_t[4]>(a)[0], ...., bitcast<int8_t[4]>(d)[3])
```

Such an intrinsic would be generally useful.

Alternatively, improving C's constant expression guarantee, e.g., to support the result of type punning being a constant expression (`union U { float a; int8_t b[4]; }{ a }.b[0]`), would also solve the underlying problem (and allow implementing a `bitcast` function in a library).

Reply:
FWIW I think its' not up to these headers to fix these problems. Adding intrinsics for the different vector types seems much simpler than adding a generic bitcast intrinsic.

Reply:
Unfortunately it turns out that clang has no way I can see of enforcing that floating point arguments fold to a constant, even for builtin functions, so I don't see a way of enforcing that f32x4 and f64x2 consts get constant arguments. Perhaps we should just document this requirement and leave the rest up to the user? Does anyone else know a way to enforce this condition?

Reply:
> Unfortunately it turns out that clang has no way I can see of enforcing that floating point arguments fold to a constant, even for builtin functions, so I don't see a way of enforcing that f32x4 and f64x2 consts get constant arguments. 

cc @zygoloid - do you know if its possible to do better here?

If we can't do better, then that's the way it is. As long as we document them, we can always enforce these constraints later.

Reply:
@tlively I think it would be ok to just add different `wasm_v128_const_..` methods for the different vectors, and to enforce the compile-time constant requirement only for the ones taking integers, if that's the limitation. For the ones taking floating-point values, you can just document that the intrinsic API has a bug that allows run-time values, but that this is not allowed by the spec.